_posts/2020-11-05-springCloudCircuitBreaker.md

```diff
@@ -0,0 +1,550 @@
+import org.springframework.retry.annotation.Recover;
+import org.springframework.retry.annotation.Retryable;
+ 
+public interface BackendAdapter {
+ 
+    @Retryable(value = { RemoteServiceNotAvailableException.class }, maxAttempts = 3, backoff = @Backoff(delay = 1000))
+    public String getBackendResponse(boolean simulateretry, boolean simulateretryfallback);
+ 
+    @Recover
+    public String getBackendResponseFallback(RuntimeException e);
+ 
+}
+
+```
+@Retryable – This is the main annotation after @EnableRetry. 
+This annotation tells that if we get RemoteServiceNotAvailableException from the method then retry maximum 3 times before sending the response. 
+Also we are introducing delay of 1 second in each retry.
+@Recover – in the fallback method indicates that if we don’t get any success response after 3 retry, response will come from this fallback method. 
+Make sure you pass expected exception as parameter, else spring will have hard time finding the exact method.
+```
+package com.example.springretry;
+ 
+import java.util.Random;
+import org.springframework.stereotype.Service;
+ 
+@Service
+public class BackendAdapterImpl implements BackendAdapter {
+ 
+    @Override
+    public String getBackendResponse(boolean simulateretry, boolean simulateretryfallback) {
+ 
+        if (simulateretry) {
+            System.out.println("Simulateretry is true, so try to simulate exception scenario.");
+ 
+            if (simulateretryfallback) {
+                throw new RemoteServiceNotAvailableException(
+                        "Don't worry!! Just Simulated for Spring-retry..Must fallback as all retry will get exception!!!");
+            }
+            int random = new Random().nextInt(4);
+ 
+            System.out.println("Random Number : " + random);
+            if (random % 2 == 0) {
+                throw new RemoteServiceNotAvailableException("Don't worry!! Just Simulated for Spring-retry..");
+            }
+        }
+ 
+        return "Hello from Remote Backend!!!";
+    }
+ 
+    @Override
+    public String getBackendResponseFallback(RuntimeException e) {
+        System.out.println("All retries completed, so Fallback method called!!!");
+        return "All retries completed, so Fallback method called!!!";
+    }
+}
+```
+
+Testing the retry methods from Spring Retry
+Start with 'http://localhost:8080/retry?simulateretry=true&simulateretryfallback=false' in your browser. Based on the parameter, we are expecting exception in the backend service call and at the same time as simulateretryfallback=false, we are depending on the random logic (random % 2 == 0 –> even random number) we can expect a success response while retry.
+So once we hit the request in browser, we might get exception in backend and spring will retry the same method multiple times. The outcome could be the Success response from backend. Here are the few lines of log from one of my request where spring is trying retry.
```

Comment:
```suggestion
Testing the retry methods from Spring Retry
Start with browsing to 'http://localhost:8080/retry?simulateretry=true&simulateretryfallback=false'. 
Based on the parameter, we are expecting exceptions and because `simulateretryfallback` is false, we are depending on the random logic (random % 2 == 0 –> even random number) that will random give us a successful response while retrying.
So once we hit the request in the browser, we might get an exception in the backend and spring will retry the same method multiple times. The outcome could be a successful response from the backend. Here are a few lines of the log from one of my requests where spring is retrying.
```